client/gutenberg/extensions/giphy/edit.js

```diff
@@ -0,0 +1,185 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { __ } from 'gutenberg/extensions/presets/jetpack/utils/i18n';
+import classNames from 'classnames';
+import { Component, createRef } from '@wordpress/element';
+import { TextControl } from '@wordpress/components';
+import { RichText } from '@wordpress/editor';
+import { debounce } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { settings } from './settings.js';
+
+class GiphyEdit extends Component {
+	timer = null;
+	textControlRef = createRef();
+
+	state = {
+		focus: false,
+	};
+
+	componentWillUnmount() {
+		this.parseSearch.cancel();
+	}
+
+	onSearchTextChange = searchText => {
+		const { setAttributes } = this.props;
+		setAttributes( { searchText } );
+		this.parseSearch( searchText );
+		this.maintainFocus();
+	};
+
+	parseSearch = debounce( searchText => {
+		let giphyID = null;
+		// If search is hardcoded Giphy URL following this pattern: https://giphy.com/embed/4ZFekt94LMhNK
+		if ( searchText.indexOf( '//giphy.com/gifs' ) !== -1 ) {
+			giphyID = this.splitAndLast( this.splitAndLast( searchText, '/' ), '-' );
+		}
+		// If search is hardcoded Giphy URL following this patterh: http://i.giphy.com/4ZFekt94LMhNK.gif
+		if ( searchText.indexOf( '//i.giphy.com' ) !== -1 ) {
+			giphyID = this.splitAndLast( searchText, '/' ).replace( '.gif', '' );
+		}
+		// https://media.giphy.com/media/gt0hYzKlMpfOg/giphy.gif
+		const match = searchText.match(
+			/http[s]?:\/\/media.giphy.com\/media\/([A-Za-z0-9\-\.]+)\/giphy.gif/
+		);
+		if ( match ) {
+			giphyID = match[ 1 ];
+		}
+		if ( giphyID ) {
+			return this.fetch( this.urlForId( giphyID ) );
+		}
+
+		return this.fetch( this.urlForSearch( searchText ) );
+	}, 250 );
+
+	urlForSearch = searchText => {
+		const { apiKey } = settings;
+		return `//api.giphy.com/v1/gifs/search?q=${ encodeURIComponent(
+			searchText
+		) }&api_key=${ apiKey }&limit=1`;
+	};
+
+	urlForId = giphyId => {
+		const { apiKey } = settings;
+		return `//api.giphy.com/v1/gifs/${ giphyId }?api_key=${ apiKey }`;
+	};
+
+	splitAndLast = ( array, delimiter ) => {
+		const split = array.split( delimiter );
+		return split[ split.length - 1 ];
+	};
+
+	fetch = url => {
+		const { setAttributes } = this.props;
+		const xhr = new XMLHttpRequest();
```

Comment:
Did a quick test with `apiFetch` and encountered a CORS error. In the test I replaced the `fetch()` function with the following:

```fetch = url => {
    const { setAttributes } = this.props;
    apiFetch( { url } ).then(
      result => {
        console.log( 'API Success', result );
      },
      result => {
        console.log( 'API Failure', result );
      }
    );
};```

This resulted in the following error in the console:

```Access to fetch at 'http://api.giphy.com/v1/gifs/search?q=ron%20swanson&api_key=LtJmY9DnDJUA0&limit=1&_locale=user' from origin 'http://localhost' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'.```

Also noting that the Gutenberg Handbook's API Fetch docs ( https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-api-fetch/ ) describe it as `Utility to make WordPress REST API requests.`. If we can make it work, I think we should, but it may be that the library wasn't intended to be used with external APIs.

Reply:
Interesting — I've understood `wp.apiFetch` to be just a thin layer over [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and kind of generic utility.

Reply:
Asked: https://wordpress.slack.com/archives/C02RQBWTW/p1548167770493500

Reply:
I encountered the same problem with the Mapbox location lookups in the Map block and ended up choosing the path of least resistance and using vanilla XMLHttpRequest: https://github.com/Automattic/wp-calypso/blob/master/client/gutenberg/extensions/map/location-search/index.js#L65-L79

If we can find resolution that would be great as using `apiFetch` is much nicer than rolling our own every time we need to communicate with a non-WP API.

Reply:
`apiFetch` as a wrapper over `window.fetch` should work fine. I'm seeing `window.fetch` requests working without any CORS issues.

I do see some issues with `apiFetch` related to some middleware it's using that is preventing external APIs —not `public-api.wordpress.com`— from being reached.

This doesn't need to block this from landing, especially as a beta block. We can make this change after it's landed and the requirements are satisfied.

Reply:
This patch mostly works using  `window.apiFetch`, `@wordpress/api-fetch` should be a bit nicer once we get over the hurdles:

```patch
diff --git a/client/gutenberg/extensions/giphy/edit.js b/client/gutenberg/extensions/giphy/edit.js
index 93bab39d28..1877a7c8d6 100644
--- a/client/gutenberg/extensions/giphy/edit.js
+++ b/client/gutenberg/extensions/giphy/edit.js
@@ -1,8 +1,7 @@
-/** @format */
-
 /**
  * External dependencies
  */
+import apiFetch from '@wordpress/api-fetch';
 import { __ } from 'gutenberg/extensions/presets/jetpack/utils/i18n';
 import classNames from 'classnames';
 import { Component, createRef } from '@wordpress/element';
@@ -77,31 +76,29 @@ class GiphyEdit extends Component {
 		return split[ split.length - 1 ];
 	};
 
-	fetch = url => {
-		const { setAttributes } = this.props;
-		const xhr = new XMLHttpRequest();
-		xhr.open( 'GET', url );
-		xhr.onload = () => {
-			if ( xhr.status === 200 ) {
-				const res = JSON.parse( xhr.responseText );
-				const giphyData = res.data.length > 0 ? res.data[ 0 ] : res.data;
-				// No results
-				if ( ! giphyData.images ) {
-					return;
-				}
-				const calculatedPaddingTop = Math.floor(
-					( giphyData.images.original.height / giphyData.images.original.width ) * 100
-				);
-				const paddingTop = `${ calculatedPaddingTop }%`;
-				const giphyUrl = giphyData.embed_url;
-				setAttributes( { giphyUrl, paddingTop } );
-				this.maintainFocus( 500 );
-			} else {
-				// Error handling TK
-			}
-		};
-		xhr.send();
-	};
+	fetch = url =>
+		window
+			.fetch( url )
+			.then( resp => resp.json() )
+			.then(
+				( { data } ) => {
+					const { setAttributes } = this.props;
+					// No results
+					const giphyData = data.length > 0 ? data[ 0 ] : data;
+					if ( ! giphyData.images ) {
+						return;
+					}
+					const calculatedPaddingTop = Math.floor(
+						( giphyData.images.original.height / giphyData.images.original.width ) * 100
+					);
+					const paddingTop = `${ calculatedPaddingTop }%`;
+					const giphyUrl = giphyData.embed_url;
+					setAttributes( { giphyUrl, paddingTop } );
+					this.maintainFocus( 500 );
+				},
+				// eslint-disable-next-line no-console
+				'production' === process.env.NODE_ENV ? err => console.error( err ) : undefined
+			);
 
 	setFocus = () => {
 		this.maintainFocus();
```

Reply:
I made the apiFetch change in #30332 and #30333. Depending on how reviews unfold that will give us some flexibility on how we land things.

This isn't a blocker for this PR.